packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1725,6 +1726,32 @@ function validateAttributes(
     }
 }
 
+function validateSlotAttribute(
+    ctx: ParserCtx,
+    parsedAttr: ParsedAttribute,
+    parentNode: ParentNode,
+    element: BaseElement
+): void {
+    const slotAttr = parsedAttr.get('slot');
+
+    if (!slotAttr) {
+        return;
+    }
+
+    const elementOrSlotAncestor = ctx.findAncestor(
+        (node) => ast.isElement(node) || ast.isSlot(node),
+        ({ current }) => current && !ast.isComponent(current) && !ast.isExternalComponent(current),
+        parentNode
+    );
+
+    if (elementOrSlotAncestor) {
```

Comment:
```suggestion
    // Warn if a `slot` attribute is on an element that isn't an immediate child of a containing LWC component or
    // `lwc:external` component. This is a case that all three of native-shadow/synthetic-shadow/light DOM will
    // simply ignore, but it's good to warn, so that developers realize that they may be making a mistake.
    // Note that, for the purposes of being considered an "immediate child," virtual elements like `for:each` and
    // `lwc:if` don't count - only rendered elements (including `<slot>`s) count.
    // Example of invalid usage: `<x-foo><div><span slot=bar></span></div></x-foo>`
    if (elementOrSlotAncestor) {
```